### DISCUSSION TITLE
Standard WikiText & Custom Markup - DOC suggestions

### DISCUSSION THREAD

**Discussion by TiddlyTweeter** [wikilabs/plugins#63] [Custom Markup]
I'm starting this Discussion to, hopefully, explore the ways we can integrate Standard WikiText with @pmario's Custom Markup (CM) system. I'll make posts about existing WikiText behaviours one-by-one and then make a few observations and raise some questions in relation to CM.

**Comment by TiddlyTweeter** [comment]
### WikiText Tables (overview)

Standard WikiText for tables is pretty amazing! Ever since I started using TW I have used it extensively. It is very sophisticated with great "cell merge", "cell align", make cell a "th", _auto creation_ of "thead" & "tfoot", application of "class", creation of "caption". It is well documented at: [Tables In WikiText](https://tiddlywiki.com/#Tables%20in%20WikiText). In brief...

![Screenshot 2021-02-15 114305](https://user-images.githubusercontent.com/28830541/107936512-0a59fa80-6f83-11eb-9bce-fa4085280e2d.jpg)

I'll comment on some interesting issues with the approach next.

**Reply by TiddlyTweeter** [reply]
### WikiText Tables - Row Classes Are Hard Coded? ...

WikiText...
```
|Header Row||h
|first row||
|second row||
```
Render...

```<table>
<thead>
<tr **class="evenRow"**><td>Header Row</td><td></td></tr>
</thead>
<tbody>
<tr **class="oddRow"**><td>first row</td><td></td></tr>
<tr **class="evenRow"**><td>second row</td><td></td></tr></tbody>
</table>
```

@pmario ...QUESTIONS...

1- Is the application of classes `class="evenRow"` & `class="oddRow"` in render _**hard-coded**_ by the parser?
Often it is useful in simple tables. BUT it is an issue when you need to format a more complex table where it is just redundant.

2- Can the behaviour be switched **_off?_**

**Reply by pmario** [reply]
The "even-odd-Row" behaviour is "hardcoded" at: https://github.com/Jermolene/TiddlyWiki5/blob/0ed32fded996826a59832d7a7555bb16c4a57864/core/modules/parsers/wikiparser/rules/table.js#L173

It can _not_ be disabled atm

**Reply by pmario** [reply]
I did some tests with CM at: https://wikilabs.github.io/editions/custom-markup/#docs%20for%20multi-line%20table%20definition

**Reply by TiddlyTweeter** [reply]
> I did some tests with CM at: https://wikilabs.github.io/editions/custom-markup/#docs%20for%20multi-line%20table%20definition

Thanks for that. I wonder how _compact_ we can make markup for tables in Custom Markup? Jeremy's achievement at that in Standard Wikitext is pretty amazing! My main issue with the standard parsing is when I want more control over column behaviours. I'll study your notes & comment back

**Reply by TiddlyTweeter** [reply]
@pmario I thought your table examples interesting. I stole a couple of ideas from them :-). In testing using CM for tables I was interested, like you, in being able to more easily edit larger data cells. In this (with the right pragma definitions) CM is definitely easier than Standard Wikitext.

In the example below I pass attributes into the `<col>`  like  `<col span="2" style="width:8rem">` Thought it might interest you. Note that `﹙ little﹚` 's default behaviour is reassigned to `<td>`. Its a full table with all segments. You can edit and break lines wherever you want. It still renders correctly, I think. For some situations that behaviour could be helpful?

```
\customise angle=TABLE _endString=«TABLE _element=table _classes="class-table" _mode=block
\customise single=CA _element=caption
\customize single=CO _element="$macrocall" $name="addColSpan" _1="cols" _2="width"
\define addColSpan(cols,width)<col span="$cols$" style="width:$width$">
\customise single=HE _endString=‹HE _element=thead
\customise single=BO _endString=‹BO _element=tbody
\customise single=FO _endString=‹FO _element=tfoot 
\customise slash=R  _element=tr
\customise little=! _element=th
\customise little   _element=td

<style>
.text-align-right, .tr {text-align: right;}
.text-align-center, .tc {text-align: center;}
.vertical-align-top, .vt {vertical-align: top;}
.vertical-align-bottom, .vb {vertical-align: bottom;}
</style>

»TABLE.class-table-2 
›CA CAPTION
›CO:1:8rem
›CO:2:6rem 
›HE /°R ﹙ Header Row﹚﹙.tc Col 2﹚﹙.tc Col 3﹚°/‹HE
›BO 
    /°R ﹙!.tr head﹚﹙ data﹚﹙ data﹚°/
    /°R ﹙!.tr.vt head﹚﹙ data
and more
  
data﹚
﹙ data 
  
and more
  
data
  
broken over lines...﹚°/
    /°R ﹙!.tr.vt head﹚﹙ data﹚﹙ data﹚°/
‹BO
›FO /°R ﹙ Footer row﹚﹙ Col 2﹚﹙ Col 3﹚°/‹FO
«TABLE
```

**Reply by AnthonyMuscio** [reply]
I like the use of CM for tables and other complex structures build in html but if one understand the rules of html tables why not use the native element names such as tr th td etc.. as your symbols? 

Even just the automatic closure of such elements is enough to make a table much easier to read. 

Tony

**Reply by pmario** [reply]
Tony, You are right. TT and I are still experimenting with CM table syntax. ... It should be "easy to read", "easy to remember" and "easy to type" ... Especially the "cell merging" and "text alignment" will be tricky. see [some examples](https://wikilabs.github.io/editions/custom-markup/#test-table-vertical-alignment:test-table-vertical-alignment%20test-table-vertical-alignment-debug%20test-table-text-alignment%20test-table-multiline-cells%20table%2FStylesheet%20%5B%5Bdocs%20for%20multi-line%20table%20definition%5D%5D) ... That's still experiments, that only cover parts of the real possibility of html tables.

**Reply by pmario** [reply]
But now we do have better "snippet handling" so users can define some predefined table layouts that they know. So they only have to "copy paste" some new columns and rows and add some text. 

The advantage of CM here (I think) is, that it should be easier to be dynamically created. ... BUT that has to be proved ;)

**Reply by TiddlyTweeter** [reply]
> I like the use of CM for tables and other complex structures build in html but if one understand the rules of html tables why not use the native element names such as tr th td etc.. as your symbols?

@AnthonyMuscio you are likely right! 

The issue for me is that as I work with CM, which actually really fits my needs for fine-control over "IN Tiddler" design, there is a necessary evolution process to derive a **_consistent_** system for **ID's** and **parameter** and **class** names.

**Comment by TiddlyTweeter** [comment]
### WikiText Blockquotes (overview)

Blockquotes in WikiText are very rich &amp; flexible. They are almost as rich in behavior as WikiText tables. TW supports both multi-line <code>&lt;&lt;&lt; ... &lt;&lt;&lt;</code> and single-line <code>&gt;</code>(+) blockquotes. The docs for <code>&lt;blockquote&gt;</code> insertion via WikiText are goodly documented at: <a class="tc-tiddlylink-external" href="https://tiddlywiki.com/#Block%20Quotes%20in%20WikiText" rel="noopener noreferrer" target="_blank">Block Quotes in WikiText</a>. In brief...

![Screenshot 2021-02-15 170207](https://user-images.githubusercontent.com/28830541/107969346-f4166380-6faf-11eb-8098-3040cc2ef6b4.jpg)

I'll comment later on an issue or two...

**Reply by TiddlyTweeter** [reply]
### WikiText Blockquotes Doesn't Insert The `cite="URL"` Attribute. Maybe it should?

@pmario one issue with `blockquote` is the `cite="URL"` ATTRIBUTE, which it doesn't insert into the element.  

In Standard WikiText we can add a `<cite> ... </cite>` element, which is good. BUT just as important would be to be able to add an _attribute_ for a cite="url" to the  `blockquote` element itself, see: https://www.w3schools.com/TAGS/tag_blockquote.asp.

Could Custom Markup pass such needed _attributes_ into elements? (_Dynamically from text, not from a `\customise` pragma._)

_Your thoughts?_

**Reply by pmario** [reply]
>..Could Custom Markup pass such needed attributes into elements? (Dynamically from text, not from a \customise pragma.)

I think it may be possible with some specialized macros. ... I think if a macro can do it, CM may be able to call it.

**Reply by TiddlyTweeter** [reply]
> > ..Could Custom Markup pass such needed attributes into elements? (Dynamically from text, not from a \customise pragma.)
> 
> I think it may be possible with some specialized macros. ... I think if a macro can do it, CM may be able to call it.

@pmario, that could be very useful in a few special cases where making a pragma just for an attribute used once is not really viable. Some of the cases where it could be good ...

>`abbr`--- `title="Full String Of Abbreviation"` --- Essential for abbreviations to work properly. They are always different!
>
>`blockquote` --- `cite="URL"`
>
>`col` --- `span="No."` --- For adjustment of table columns.
>
>`div` --- `style="rule: value;"` --- For fine adjustment of divisions (for instance width for poems etc.)

**Reply by pmario** [reply]
This is a blockqoute definition with an URL parameter BUT the URL can't use dots eg: `test.com` won't work atm

```
\customize degree=blockquote _element="$macrocall" $name="test" _1="url" _2="citeTxt" _mode=block _endString="---"

\define test(src, url, citeTxt)
<blockquote>
$src$
<cite url="http://$url$">$citeTxt$</cite>
</blockquote>
\end

°blockquote:url-with-no-spaces-and-no-dots:cite-text
line1

line2

line3
---
```

**Reply by pmario** [reply]
I'll probably have to adjust the parser once more

**Reply by pmario** [reply]
For me personally the cite text including the link makes much more sense than the "cite" parameter, which is only machine readable and doesn't make sense for humans. 

Have a closer look at: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/blockquote and https://developer.mozilla.org/en-US/docs/Web/HTML/Element/cite which will also show you some links to the spec. 

```
<<<
some text to quote
<<<[[external link|https://tiddlywiki.com]]
```

**Reply by pmario** [reply]
This example **does not work with 0.9.0**. It will start to work with 0.10.0

I did adjust the parser, so it's possible to pass on parameters, that contain spaces. That's an improvement, but it will also be a "breaking change". 

AND

It's an example, how to use _1 and _2 params.

```
\customize degree=blockquote _element="$macrocall" $name="test" _1="citeTxt" _2="url" _mode=block _endString="---"

\define test(src, citeTxt, url)
<blockquote cite="http://$url$">
$src$
<cite><a class="tc-tiddlylink-external" href="https://$url$" rel="noopener noreferrer" target="_blank">$citeTxt$</a></cite>
</blockquote>
\end

°blockquote:"link with spaces":"tiddlywiki.com"
line1

line2

line3
---

------

<<<
some quote text
<<<[[external link|https://tiddlywiki.com]]
```

The second thing you can see, if you view "HTML raw" output is, that wikitext does a lot of adjustments to the html code, which are important. eg: `rel="noopener noreferrer" target=` and so on. 

It's very likely, that these elements are forgotten by users. Some of those adjustments are needed to protect privacy and security. ... So CM is powerfull, but you have to know, what you do.

**Reply by pmario** [reply]
I'll upload the new version soon

**Reply by TiddlyTweeter** [reply]
Thanks very much @pmario! CM is great. I was able to work out from your examples how to do `<abbr title="value">` . Here is my attempt at that ...

```
\customize slash=abbr _element="$macrocall" $name="addAttrib" _1="value" _2="addText"
\define addAttrib(value,addText)<abbr title="$value$">$addText$</abbr>

For some time I have been using /°abbr:TiddlyWiki:TW °/ to write with.

Nowadays we all use the /°abbr:World-Wide-Web:WWW °/.
```

**Reply by TiddlyTweeter** [reply]
> I did adjust the parser, so it's possible to pass on parameters, that contain spaces.

That is great news! I noticed the issue.

**Reply by TiddlyTweeter** [reply]
> For me personally the cite text including the link makes much more sense than the "cite" parameter, which is only machine readable and doesn't make sense for humans.

Right. I need it only for my use. I periodically run regex over my wikis and extract all urls for checking and other purposes.

**Reply by TiddlyTweeter** [reply]
### Example: `<q>` element, `lang=` attribute...

Here is another example (proof of concept) that uses the `lang="[country code]"` for the inline `<q>` element. The attribute changes the style of quote marks to match the language ...
```
\customize slash=Q _element="$macrocall" $name="addAttribs" _1=p1 _2=p2
\define addAttribs(p1,p2)<q $p1$>$p2$</q>

<style>q {font-size: 1.1em; font-style: italic;}</style>

As they say in English /°Q:lang="en":foot-in-the-mouth °/ is when you make a gaffe, a mistake, especially when talking. In German /°Q:lang="de":ins-Fettnäpfchen-treten °/ is similar, as is the Italian /°Q:lang="it":fare-una-gaffe °/.
```
Output...
![Screenshot 2021-02-17 082249](https://user-images.githubusercontent.com/28830541/108169922-ac035800-70f9-11eb-9dc6-bbfc511f763e.jpg)

**Reply by TiddlyTweeter** [reply]
### Note On Example For Paramater Passing To Blockquotes

> @pmario  This example **does not work with 0.9.0**. It will start to work with 0.10.0
> I did adjust the parser, so it's possible to pass on parameters, that contain spaces. 
> ... It's an example, how to use _1 and _2 params.
> 
> ```
> \customize degree=blockquote _element="$macrocall" $name="test" _1="citeTxt" _2="url" _mode=block _endString="---"
> 
> \define test(src, citeTxt, url)
> <blockquote cite="http://$url$">
> **$src$**
> <cite><a class="tc-tiddlylink-external" href="https://$url$" rel="noopener noreferrer" target="_blank">$citeTxt$</a></cite>
> </blockquote>
> \end

Very good example! Thanks! I now understand it! But I did _not_ understand at first that **$src$** is special! :-) 
Basically it is the CONTENT between the opening and closing Glyphs? Right? It is captured by your parser, right? 
It is hard-coded somewhere?

Anyway, I now can use parameter passing pretty well!

 Q: @pmario BUT, in final docs, it may be worth explaining what **$src$** does? Something like that.

**Reply by pmario** [reply]
>Very good example! Thanks! I now understand it! But I did not understand at first that $src$ is special! :-)

Yea, The `src` parameter was the first one, that I did introduce.

>Basically it is the CONTENT between the opening and closing Glyphs? Right? 

Exactly! ... It was the easiest one to implement. Almost no additional code was needed. No config, no pragma settings. .. 

>It is captured by your parser, right?

yes

>It is hard-coded somewhere?

yes. It will be always there, even if you don't need it ;)

**Reply by TiddlyTweeter** [reply]
TX! Great to know! And WORKS well!

> > It is hard-coded somewhere?
> 
> yes. It will be always there, even if you don't need it ;)

:-) I just think that in docs that should be mentioned. That is all!

**Reply by AnthonyMuscio** [reply]
TT,

Just reviewed this discussion thread now. You talk about including a `<cite>` parameter `URL=` in the wikitext markup for block quotes. This may make sense in a "custom markup free environment", but with custom markup could you swap this around?. Add block quote css to custom markup for the html cite element?

It seems to me we should embrace HTML and all its features and unleash them with custom markup, rather than add individual features to wikitext markup. One starts with many features from html where you select elements for custom wikitext, the other is custom wikitext to which you add additional features. The former approach is the most expansive in my view, html is universal and maintained for a much larger community; the latter wiki markup is smaller and maintained by a smaller community.

Your thoughts?

**Comment by pmario** [comment]
The cool thing after a pause using and developing CM is, that you can see, what's missing and which elements still have rough edges

**Comment by TiddlyTweeter** [comment]
### Issues Arising - Just My Thoughts

(This is horribly OPINIONATED! :-)

As I work more with Custom Markup I realise, more and more, that the basic ideas behind it share a lot with Standard WikiText.
And it is making me aware that the potential of Standard WikiText is likely quite _under-used._ _Why?_ I don't think there is general wide understanding of how to pass styles into WikiText constructs. Partly this is because there are several methods...

- Passing in `class=` attribute via `.class` appended to the WikiText construct
- Wrapping in class `@@.class ... @@` via `<span>` (inline)
- Passing in `class=` attribute via `@@.class ... @@` (some blocks)
- Passing in to the `style=` attribute `@@property:value; ... @@`  

BUT there is an _issue_ in that the way they behave is _not fully consistent_. They are _context dependent_. 
But we LACK, I think, full enough docs to fully understand _WHICH method to use where?_

_IMO_, Custom Markup opens the way to fully use HTML & CSS in TW in an efficient way IF a writer needs to quickly customise content.
But it also IMPROVES users understanding of Standard WikiText!

QUESTION ...

@pmario _Should the docs for Custom Markup ALSO explain Standard WikiText better too?_

(P.S. My thinking here is you can ALSO use Custom Markup to QUICKLY prototype new CSS use. It could be used well as part of the development of new CSS for "Standard" use in TW??)

**Reply by pmario** [reply]
>BUT there is an issue in that the way they behave is not fully consistent. They are context dependent.

That's true for many TW features. They have been developed "on demand", depending on bug-reports and user-requests in the GG. That's also the reason, why _not all_ widgets have the exact same behaviour and parameters, that work in the exact same way. 

We tried to fix the "missing feature" problems with the latest releases. But there may be consistency problems that have to stay! 
**Because: Backwards compatibility is more important!!** 

>IMO, Custom Markup opens the way to fully use HTML & CSS in TW in an efficient way IF a writer needs to quickly customise content.
But it also IMPROVES users understanding of Standard WikiText!

Yes. That's true. IMO The deeper you dig into manipulating TW "custom" syntax, the more you start to understand the internals. custom markup only works **inside** the existing TW "standard" rules. We are just "bending" them by making more "parameters" available. 

Eg: The standard wikitext "_endString" is hardcoded and always produces a paragraph. Now we can change this behaviour.

**Reply by pmario** [reply]
>QUESTION ...
>
>@pmario Should the docs for Custom Markup ALSO explain Standard WikiText better too?
>
>(P.S. My thinking here is you can ALSO use Custom Markup to QUICKLY prototype new CSS use. It could be used well as part of the development of new CSS for "Standard" use in TW??)

I think, if something for the "standard" doc is needed to understand custom-markup, it should go to tiddlywiki.com standard docs. CM docs should only link there. 

IMO This will help both projects.

**Reply by TiddlyTweeter** [reply]
> @pmario  there may be consistency problems that have to stay!
> **Because: Backwards compatibility is more important!!**

Right! One of the biggest appreciations of TW I have is, basically, its serious commitment to _backwards compatibility_.

BUT the issue for an end user is that the various changes over time have led to some inconsistency that can start becoming a _"barrier-to-entry"_ too?

On WikiText ORTHOGRAPHY for styling, in particular, the inconsistencies (simply on what you "enter") can be pretty confusing, I think?

**Reply by pmario** [reply]
>BUT the issue for an end user is that the various changes over time have led to some inconsistency that can start becoming a "barrier-to-entry" too?
>
>On WikiText ORTHOGRAPHY for styling, in particular, the inconsistencies (simply on what you "enter") can be pretty confusing, I think?

You are right. That's why we do have a "deprecated" mechanism. .. The problem is, that this mechanism isn't really used. ... If we deprecate something we would be able to move it into a plugin, after 1 or 2 years. ... BUT we actually never did that.

**Reply by TiddlyTweeter** [reply]
> @pmario ... You are right. That's why we do have a "deprecated" mechanism. .. The problem is, that this mechanism isn't really used. ... If we deprecate something we would be able to move it into a plugin, after 1 or 2 years. ... BUT we actually never did that.

Ah! Okay. I just want to comment that in order to fully appreciate Custom Markup plugin I _did_ need to better understand the "(semi-)deprecated" existing features of TW! :-)

**Reply by AnthonyMuscio** [reply]
Mario;

You say;
> Eg: The standard wikitext "_endString" is hardcoded and always produces a paragraph. Now we can change this behaviour.

but does not the line wikitext like * # produce line feeds? as normal `<li>` elements do? ie the html rules?

Tones

**Comment by TiddlyTweeter** [comment]
### Adding Custom Pragma That Compliment Standard WikiText? - No. 1

In Standard WikiText we have the constructs for...
-   [via &#x0060;&#x0060;&#x0060; ... &#x0060;&#x0060;&#x0060;] `<pre><code> ... </code></pre>` Docs: [Code Blocks in WikiText](https://tiddlywiki.com/#Code%20Blocks%20in%20WikiText)
and ...
- [via `""" ... """`] `<p> ... <br> etc ...</p>` Docs: [Hard Linebreaks in WikiText](https://tiddlywiki.com/#Hard%20Linebreaks%20in%20WikiText) 

I thought it interesting to add ... Simple PRE ...
```
\customise tick="'''" _endString="'''´" _element=pre _classes="test-style"

´''' test

  enter `´'''` to start `<pre>` block

  enter `'''´` to end `</pre>` block

test
'''´
```
... render ...

```
<pre class="test-style wltc-l1 wltc">test

  enter <code>´'''</code> to start <code>&lt;/pre&gt;</code> block

  enter <code>'''´</code> to end <code>&lt;/pre&gt;</code> block

test
</pre>
```

**Reply by pmario** [reply]
uuuh. That's an interesting one. You really have to test it, to see, what's actually going on. cool!

**Reply by TiddlyTweeter** [reply]
It needs CSS-style cascade change for "`pre > code`" elements to display. Normal TW CSS `pre > code` render the same.

**Comment by TiddlyTweeter** [comment]
### Adding Custom Pragma That Compliment Standard WikiText? - No. 2

This one adds a `<pre>` container to emulate a computer CONSOLE :-). Useful in documentation of Windows command prompt ... :-)

I will improve it by PASSING in ATTRIBUTES on style, so the user can change the width and positioning in text flow. Later!

```
\rules except dash
\customise angle=CONSOLE _endString="«CONSOLE" _element=pre style="max-width:31rem;height:100%;font-family:monospace;font-size:.8rem;line-height:1;color:yellow;font-weight:600;background-color:black;border:3px solid grey;padding:1rem;border-radius:.8rem;"

<style>.blink-console {background-color: white; color:transparent; animation:blinking 1.4s infinite;}
@keyframes blinking{0%{background-color: white;}47%{background-color: transparent;}}
</style>

»CONSOLE C:\Users\TT>

               Powershell 5.1.19041.610

             +===========================+
              FETCH TW - OWLET v01a
            +=============================+
            |    TW5 RELEASES             |
            |  D - Release: With Docs     |
            |  E - Release: Empty         |
            |  F - Release: Full Plugins  |
            |      ---------------------  |
            | PD - Prerelease: With Docs  |
            | PE - Prerelease: Empty      |
            |      ---------------------  |
            |  U - Upgrade Tool           |
            | --------------------------- |
            |    TW5 EDITIONS             |
            | E1 - CodeMirror             |
            | E2 - D3 Graphics            |
            | E3 - Highlights             |
            | E4 - KaTex                  |
            | E5 - Markdown               |
            | E6 - Resume Builder         |
            | E7 - Text Slicer            |
            | E8 - XLSX Utilities         |
            | --------------------------- |
            |    TWC RELEASES             |
            |  C - Release                |
            | --------------------------- |
            | H - Help                    |
            | --------------------------- |
            | X - Exit                    |
            +=============================+
              Internal Settings active
             +===========================+
?             key:/°.blink-console _°/

«CONSOLE
```
 ... outputs ...

![Screenshot 2021-02-22 160153](https://user-images.githubusercontent.com/28830541/108726082-54188700-7527-11eb-9414-bcff17c055be.jpg)

**Reply by AnthonyMuscio** [reply]
Really cool

One of possibly hundreds of canned "custom wikitext" authors would like to have up their sleeves for producing rich content.

**Reply by TiddlyTweeter** [reply]
> @AnthonyMuscio One of possibly hundreds of canned "custom wikitext" authors would like to have up their sleeves for producing rich content.

Right! I think it is quite a striking example, though it was actually easy to create and is easy to use. You can see the latest full version online at: [Console Emulation Example: Parameter Passing](https://tidbits.wiki/lab/cm/cm-tech.html#%22Console%22%20Block%2C%20With%20Dynamic%20Parameter%20Settings)

**Comment by TiddlyTweeter** [comment]
### One thing that Custom Markup can ADD to Standard WikiText is _DYNAMIC_ `<a>` attributes ...

What do I mean? Merely that the brilliant `[[title of link|link]]` method in Standard TW is not always flexible enough ... This illustrates a solution via being about to invoke a site in many ways (e.g. YOUR site :-) through CM ...

![Screenshot 2021-03-07 115306](https://user-images.githubusercontent.com/28830541/110237343-dba8c180-7f3b-11eb-851d-b530cb62772c.jpg)

**Comment by TiddlyTweeter** [comment]
### Easier Way to do a Modal???

I think Modals are underused? They are great for reference materials.

In experiments with Custom Markup I found it was easier (once you done the macro) to invoke a modal 
... Here is an example, which could be improved, in time, by passing in attributes for styling.

![Screenshot 2021-03-10 101737](https://user-images.githubusercontent.com/28830541/110606527-8e6e5f00-818a-11eb-9335-d50c5a1a5624.jpg)

The button for it ...
![Screenshot 2021-03-10 103032](https://user-images.githubusercontent.com/28830541/110607731-c7f39a00-818b-11eb-841a-39d2dc1ce064.jpg)


The code a user would enter is only `⠒HLP:"Custom Markup Defaults" $:/tt/cm/hlp_Custom Markup Defaults⠶`

**Reply by pmario** [reply]
I personally hate dropdowns and modals. I think they waste screen space, and animated modals waste time too.

... If there is a different way to show additional content without them I prefer the alternative way. 

But I'm open to learn new ways to use them. ...

**Reply by pmario** [reply]
I was thinking about the `wltc-lx` classes. I think I can remove them from glyphs, that can't do indents. ... We'll see

**Reply by TiddlyTweeter** [reply]
> I was thinking about the `wltc-lx` classes. I think I can remove them from glyphs, that can't do indents. ... We'll see

Thanks. TBH, classes that don't do anything are more of a maintenance headache than they are worth? I do think the indents `wltc-l2` and up may well be used by some users wanting an easy solution to indenting so maybe worth keeping?

My 2 cents.

**Reply by TiddlyTweeter** [reply]
> I personally hate dropdowns and modals. I think they waste screen space, and animated modals waste time too.
> 
> ... If there is a different way to show additional content without them I prefer the alternative way.
> 
> But I'm open to learn new ways to use them. ...

Interesting! _I broadly agree with you._ Overuse of them is just a mess. BUT Pop-up or Modal _simple help info_ I, myself, find useful for quick reference.

**Reply by AnthonyMuscio** [reply]
Question?

Is there scope and end options for "Special Block Defaults"?

Tony

**Reply by TiddlyTweeter** [reply]
In my tests _I don't think you can redefine_ the  behavior of "Special Block Defaults" (e.g. `»»» `) except through style sheets.

As far as I can see @pmario intended them as hard-coded support for common needs for indenting, and provides classes to define what they do.

TT

**Comment by TiddlyTweeter** [comment]
### Comment On Custom Markup & Standard Markup CSS Application

I have been looking at [the numerous ways TW permits CSS styling](https://tiddlywiki.com/#How%20to%20apply%20custom%20styles). I was pleasantly surprised to see Custom Markup also supports the `@@style:value; @@` wrap method both for inline and blocks. For instance ...

> @@background-color: red;
> °.a-style.another BLOCK TEST
> @@
> 
> Inline @@background-color:red;『.a-style.another TEST』@@ of styling.

... producing ...

![Screenshot 2021-03-12 150453](https://user-images.githubusercontent.com/28830541/110950614-57dc4400-8344-11eb-97b6-a34c625072f4.jpg)

... with HTML ...

```
<div class="a-style another wltc-l1 wltc" style="background-color: red;">BLOCK TEST</div>
<p>Inline <span class="tc-inline-style" style="background-color:red;"><span class="a-style another wltc-inline wltc">TEST</span></span> of styling.</p>
```

Its handy for rare situations? _Q: Is it mentioned in Docs?_

**Reply by AnthonyMuscio** [reply]
TT,

Inline @@background-color:red; TEST@@ of styling. 

works, what is `『.a-style.another TEXT』` doing?

Tony

**Reply by pmario** [reply]
`『.a-style.another TEXT』` It assigns 2 classes: `class="a-style another"` to the SPAN element. ... The html element that is used depends on the custom pragma.

It's the same mechanism as: 

```
*.test asdf
```

that is part of TW already. This mechanism works for custom block and custom inline definitions
